taf/testlib/linux/etcd_helper.py

```
@@ -0,0 +1,118 @@
+
+
+class EtcdHelperException(Exception):
+    pass
+
+
+class EtcdHelper(object):
+
+    CLASS_LOGGER = loggers.ClassLogger()
+
+    def __init__(self, endpoint):
+        if isinstance(endpoint, str):
+            etcd_protocol, address_port = endpoint.split('://')
+            etcd_address, etcd_port = address_port.split(':')
+            self.etcd_config = {
+                'host': etcd_address,
+                'port': int(etcd_port),
+                'protocol': etcd_protocol
+            }
+        elif isinstance(endpoint, dict):
+            self.etcd_config = endpoint
+
+        self.etcd = etcd.Client(**self.etcd_config)
+
+        self._root_key = ROOT_KEY
+        self._cwd = ROOT_KEY
+        self._latest_id_key = '/'.join([ROOT_KEY, 'latest'])
+
+    def init_etcd(self):
+        self.CLASS_LOGGER.debug("Initializing etcd test entries")
+        self.etcd.write(self._latest_id_key, "0")
+
+    def change_dir(self, directory):
+        self._cwd = '/'.join([self._root_key, directory])
+
+    def _get_key(self, item):
+        return '/'.join(itertools.chain([self._cwd], item.split('__')[1:]))
+
+    def _get_root_key(self, item):
+        return '/'.join(itertools.chain([self._root_key], item.split('__')[1:]))
+
+    def __getattr__(self, item):
+        if item.startswith('key__'):
+            return self._get_key(item)
+        elif item.startswith('rootvalue__'):
+            return self.etcd.read(self._get_root_key(item))
+        elif item.startswith('value__'):
+            return self.etcd.read(self._get_key(item))
+        raise AttributeError('Unknown attribute {}'.format(item))
+
+    def __setattr__(self, item, value):
+        if item.startswith('rootvalue__'):
+            self.etcd.write(self._get_root_key(item), value)
+        elif item.startswith('value__'):
+            self.etcd.write(self._get_key(item), value)
+        else:
+            super().__setattr__(item, value)
+
+    @property
+    def latest_id(self):
+        try:
```

Please use suppress:

```
with suppress(AttributeError):
  return self._latest_id
with suppress(Exception):
  for _ in range(2):
    with suppress(etcd.EtcdKeyNotFound):
      self._latest_id = ...
      return self._latest_id
    self.init_etcd()
raise EtcdHelperException(...)
```

---

why are we suppressing Exception?  I thought we implemented something that had an exclude list so we don't suppress SyntaxError, NameError, etc.  

No code should ever catch just Exception.

---

I used the Mac's suppress version leaving out the supress(Exception). If something unknown happens it should drop down to the caller and the exception call stack should not be overwritten with calling raise.